pkg/runtime/runtime_test.go

```diff
@@ -5421,30 +5421,20 @@ func TestGracefulShutdownPubSub(t *testing.T) {
 	mockAppChannel := new(channelt.MockAppChannel)
 	rt.appChannel = mockAppChannel
 	mockAppChannel.On("InvokeMethod", mock.MatchedBy(matchContextInterface), matchDaprRequestMethod("dapr/subscribe")).Return(fakeResp, nil)
-
 	require.NoError(t, rt.initPubSub(cPubSub))
 	mockPubSub.AssertCalled(t, "Init", mock.Anything)
 	rt.startSubscriptions()
 	mockPubSub.AssertCalled(t, "Subscribe", mock.AnythingOfType("pubsub.SubscribeRequest"), mock.AnythingOfType("pubsub.Handler"))
-	assert.NotNil(t, rt.pubsubCtx)
-	assert.NotNil(t, rt.topicCtxCancels)
-	assert.NotNil(t, rt.topicRoutes)
+	assert.Nil(t, rt.pubsubCtx.Err())
 	rt.running.Store(true)
 	go sendSigterm(rt)
-	select {
```

Comment:
Rather than deleting this and adding a timeout, which makes tests slower, how about you run this `select` before invoking `sendSigterm`, in a goroutine?

```go
rt.runtimeConfig.GracefulShutdownDuration = 2 * time.Second
	testDoneCh := make(chan struct{}, 1)
	doneCh := rt.pubsubCtx.Done()
	go func() {
		select {
		case <-doneCh:
			// nop
		case <-time.After(rt.runtimeConfig.GracefulShutdownDuration + 2*time.Second):
			assert.Fail(t, "pubsub shutdown timed out")
		}
		testDoneCh <- struct{}{}
	}()
	sendSigterm(rt)
	<-testDoneCh
	assert.Nil(t, rt.pubsubCtx)
	assert.Nil(t, rt.topicCtxCancels)
	assert.Nil(t, rt.topicRoutes)
```

This passes with `go test -race`

Reply:
[log.txt](https://github.com/dapr/dapr/files/10403210/log.txt)

Attached log for go test -race

Reply:
@ItalyPaleAle - in runtime.go the pubsubctx is being set to nil -
https://github.com/dapr/dapr/blob/44dd89240e2c103a7f2e5ff5c768f89826168f7d/pkg/runtime/runtime.go#L3081-L3094

From blame I see that you have set pubsubCtx to nil, which is causing the race in the test. Any specific reason why the context needs to be set to nil? Can we rather use pubsubCtx.Err() to check if context is active or not?

Reply:
A bit of digging helped understand that stopSubscriptions is called irrespective of whether the subscriptions have been initialised or not. hence the nil check helps conditionally invoke cancel. If we could replace
```code
	if a.pubsubCancel != nil {
	 	a.pubsubCancel()
	 }
         pubsubCtx = nil
         pubsubCancel = nil
```  
with 
```code
	if a.pubsubCtx == nil {
		return
	}
	if a.pubsubCtx.Err() == nil {
		a.pubsubCancel()
	}
        . . .
        . . .
```
we could mitigate the need to set pubsubctx to nil. This will eliminate race  condition in any code that uses the pubsubctx variable. WDYT?
Few assumptions : 
- Context can not be revived after cancelling. So not setting this to nil does not impact functionally
- As soon as the context is cancelled, all child contexts will be cancelled 
- every usage of the context should check for pubsubctx.Err() == nil to ensure context is active and not depend on nil check